packages/lib-utils/src/k8s/k8s-utils.ts

```diff
@@ -50,3 +62,189 @@ export const getK8sResourceURL = (
+    .sort()
+    .forEach((k) => {
+      const value = matchLabels?.[k] as string;
+      requirements.push(createEquals(k, value));
+    });
+
+  matchExpressions?.forEach((me) => {
+    requirements.push(me);
+  });
+
+  return requirements;
+};
+
+export const selectorToString = (selector: Selector): string => {
+  const requirements = toRequirements(selector);
+  return requirements.map(requirementToString).join(',');
+};
+
+export const k8sWatch = (
+  kind: K8sModelCommon,
+  query: {
+    labelSelector?: Selector;
+    resourceVersion?: string;
+    ns?: string;
+    fieldSelector?: string;
+  } = {},
+  wsOptions: {
+    [key: string]: unknown;
+  } = {},
+) => {
+  const queryParams: QueryParams = { watch: 'true' };
+  const opts: {
+    queryParams: QueryParams;
+    ns?: string;
+  } = { queryParams };
+  const wsOptionsUpdated = {
+    path: '',
+    host: 'auto',
+    reconnect: true,
+    jsonParse: true,
+    bufferFlushInterval: 500,
+    bufferMax: 1000,
+    ...wsOptions,
+  };
+
+  const { labelSelector } = query;
+  if (labelSelector) {
+    const encodedSelector = encodeURIComponent(selectorToString(labelSelector));
+    if (encodedSelector) {
+      queryParams.labelSelector = encodedSelector;
+    }
+  }
+
+  if (query.fieldSelector) {
+    queryParams.fieldSelector = encodeURIComponent(query.fieldSelector);
+  }
+
+  if (query.ns) {
+    opts.ns = query.ns;
+  }
+
+  if (query.resourceVersion) {
+    queryParams.resourceVersion = encodeURIComponent(query.resourceVersion);
+  }
+
+  const path = getK8sResourceURL(kind, undefined, opts);
+  wsOptionsUpdated.path = path;
+
+  // TODO utilize ws-factory
+  return { path, wsOptionsUpdated };
+  // return new WSFactory(path, wsOptionsUpdated as WSOptions);
+};
+
+/**
+ * Provides a group, version, and kind for a k8s model.
+ * @param model k8s model
+ * @returns The group, version, kind for the provided model.
+ * If the model does not have an apiGroup, group "core" will be returned.
+ * */
+export const getGroupVersionKindForModel: GetGroupVersionKindForModel = ({
+  apiGroup,
+  apiVersion,
+  kind,
+}) => ({
+  ...(apiGroup && { group: apiGroup }),
+  apiVersion,
+  kind,
+});
+
+/**
+ * @deprecated - This will become obsolete when we move away from K8sResourceKindReference to K8sGroupVersionKind
+ * Provides a reference string that uniquely identifies the group, version, and kind of a k8s resource.
+ * @param K8sGroupVersionKind Pass K8sGroupVersionKind which will have group, version, and kind of a k8s resource.
+ * @param K8sGroupVersionKind.apiGroup Pass group of k8s resource or model.
+ * @param K8sGroupVersionKind.apiVersion Pass version of k8s resource or model.
+ * @param K8sGroupVersionKind.kind Pass kind of k8s resource or model.
+ * @return The reference for any k8s resource i.e `group~version~kind`.
+ * If the group will not be present then "core" will be returned as part of the group in reference.
+ * * */
+export const getReference = ({
```

Comment:
Debating if it's worth stripping the `~` logic out right now... It's adding complexity and I've had to waltz around it. One question I have is how will pushing and retrieving from the `Set<>` here: https://github.com/openshift/dynamic-plugin-sdk/pull/37/files#diff-8e891455963ea25676398e13f0e91cf63350df91315a520dd470da3039fd3c23R121 will handle the `{ apiGroup, apiVersion, kind }` object

Reply:
Yeah, this is a painful migration for us... we should try to make a stance on it though.